graphite/graphite.go

```diff
@@ -0,0 +1,330 @@
+package graphite
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/influxdb/influxdb"
+
+	log "code.google.com/p/log4go"
+)
+
+var (
+	// ErrBindAddressRequired is returned when starting the Server
+	// without a TCP or UDP listening address.
+	ErrBindAddressRequired = errors.New("bind address required")
+
+	// ErrServerClosed return when closing an already closed graphite server.
+	ErrServerClosed = errors.New("server already closed")
+
+	// ErrDatabaseNotSpecified retuned when no database was specified in the config file
+	ErrDatabaseNotSpecified = errors.New("database was not specified in config")
+
+	// ErrServerNotSpecified returned when Server is not specified.
+	ErrServerNotSpecified = errors.New("server not present")
+)
+
+// Graphite Server provides a tcp and/or udp listener that you can
+// use to ingest metrics into influxdb via the graphite protocol.  it
+// behaves as a carbon daemon, except:
+//
+// no rounding of timestamps to the nearest interval.  Upon ingestion
+// of multiple datapoints for a given key within the same interval
+// (possibly but not necessarily the same timestamp), graphite would
+// use one (the latest received) value with a rounded timestamp
+// representing that interval.  We store values for every timestamp we
+// receive (only the latest value for a given metric-timestamp pair)
+// so it's up to the user to feed the data in proper intervals (and
+// use round intervals if you plan to rely on that)
+type Server struct {
+	mu   sync.Mutex
+	wg   sync.WaitGroup
+	done chan struct{} // close notification
+
+	Server interface {
+		WriteSeries(database, retentionPolicy, name string, tags map[string]string, timestamp time.Time, values map[string]interface{}) error
+		DefaultRetentionPolicy(database string) (*influxdb.RetentionPolicy, error)
+	}
+
+	// The TCP address to listen on.
+	TCPAddr *net.TCPAddr
+
+	// The UDP address to listen on.
+	UDPAddr *net.UDPAddr
+
+	// The name of the database to insert data into.
+	Database string
+}
+
+// ListenAndServe opens TCP (and optionally a UDP) socket to listen for messages.
```

Comment:
They can specify one or the other (tcp or udp).  So they can just open a udp if in the config they only specify the udp protocol.